### PR TITLE
Move assertj from BOM to build-parent

### PR DESCRIPTION
**Pull request by famod**
https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/assertj-core.20in.20BOM

After this is merged, I will have a look at those failures with 3.20.1 (https://github.com/quarkusio/quarkus/pull/17968#issuecomment-863582001).

### CODE REVIEW

**Comment by gastaldi** [comment]
This will break existing applications using assertj, like https://github.com/quarkusio/registry.quarkus.io/blob/main/pom.xml#L112

Not a big deal though, as you can always add the assertj version but I guess it is worth adding something in the migration guide

**Comment by famod** [comment]
> it is worth adding something in the migration guide

:+1: will do once (if) this gets merged.

**Comment by gastaldi** [comment]
The commit message is a bit weird, rebasing issues?

**Comment by famod** [comment]
Thanks for the pointer, fixed! It's just too hot to think straight...

File: build-parent/pom.xml
Comment (gastaldi) [inline build-parent/pom.xml:290]: This needs to be declared in test scope

Comment (famod) [reply build-parent/pom.xml:290]: Well, it's only in `dependencyManagement` and I usually dont' use scopes in depMgmt but since other deps here do use test scope, I will of course add it. Again, thanks for pointing that out.

**Review by gastaldi — DISMISSED** [review]
LGTM

**Review by gsmet — CHANGES_REQUESTED** [review]
I'm not sure I like this change.

It will break all the applications relying on our BOM to provide the version and AFAICS it doesn't solve any issue per se.

I would rather add it to Dependabot and keep it updated and consider it a feature given how common this library is for testing.

**Comment by famod** [comment]
My main point is: How do users know which deps  from BOM are safe to update individually?
AFAICS, there is no "technical" reason to have it in the BOM, it's more a convenience thing.

Btw, some users stay on older Quarkus versions for a long time.

**Comment by gsmet** [comment]
Well, I can understand your point but I'm really not sure it's worth breaking a ton of apps for that.

**Comment by gastaldi** [comment]
> Well, I can understand your point but I'm really not sure it's worth breaking a ton of apps for that.

I'm not sure that is necessarily true. Since this is a major version bump, we are (supposed) to be allowed to break some things as long as we provide the necessary migration steps in the migration guide :)

**Comment by famod** [comment]
@gastaldi AFAICS, the `test-framwork` bits are _not_ end-user facing, or are they? What about the test BOM instead?

What's your take on this @aloubyansky?

I have no hard feelings just closing this PR, as I personally now know that I can safely bump it in my project, but it feels a little weird to me. 🤷

**Comment by aloubyansky** [comment]
There are parts of the test-framework that are user-facing. However, afaics, having assertj in the quarkus-bom isn't justified. But @gsmet also has a point.
I take your complaint about this @famod as a good thing :) If we could get more votes, that could help.

**Comment by famod** [comment]
> There are parts of the test-framework that are user-facing.

Yes, sure. To be clear, I meant the parts that are using assertj

**Comment by famod** [comment]
FWIW, `awaitility` is a similar case.

How to proceed?

**Comment by gastaldi** [comment]
I'd suggest to leave as is, to avoid breaking apps that reuse it from the quarkus-bom, but that's debatable

**Comment by aloubyansky** [comment]
IMO, as long as we do not require any specific version, we should remove them __at some point__.

**Comment by gastaldi** [comment]
What should we do with this PR? Close it and revisit later?

**Review by aloubyansky — APPROVED** [review]

**Comment by famod** [comment]
Friendly reminder @gsmet ⬆️

**Comment by aloubyansky** [comment]
Guillaume is on PTO until next week.

**Comment by aloubyansky** [comment]
We can move it to the test bom.

**Comment by aloubyansky** [comment]
Only the build-parent will be ok then. Thanks!

**Comment by famod** [comment]
> Only the build-parent will be ok then. Thanks!

Alright, done!

**Review by aloubyansky — APPROVED** [review]

**Comment by aloubyansky** [comment]
we need @gsmet to approve it

**Comment by famod** [comment]
> we need @gsmet to approve it

@gsmet yay or nay? Please close this PR if the latter.

**Comment by famod** [comment]
_This PR will terminate itself on its 6 months anniversary (2021-12-18)!_ 😉

**Comment by gsmet** [comment]
Sorry, I agree it would probably be better but it will break a lot of apps/extensions.

Maybe let's do that for a major release.

**Comment by famod** [comment]
_Blows dust away._ Let's see...

**Comment by gsmet** [comment]
Question: should we do the same for REST Assured?

/cc @aloubyansky

**Comment by famod** [comment]
Hum...that went wrong. Let me fix that...

**Comment by aloubyansky** [comment]
> Question: should we do the same for REST Assured?
> 
> /cc @aloubyansky

This will be more complicated. We add REST Assured as an application test dependency when we generate projects and we assume it's managed, so we add w/o a version. If we move it out, we'll have to be adding a specific version to the templates/generator and keep it up-to-date.
The good thing is, unlike assertj, I don't recall any issues having it in the quarkus-bom, did anybody hear any?

**Comment by famod** [comment]
@aloubyansky you are summoning daemons: https://github.com/quarkusio/quarkus/pull/23114#issuecomment-1019550182 😉

**Comment by aloubyansky** [comment]
If the transitive dependencies of the new version aren't compatible with the quarkus-bom then it's not a bad thing we are including a version that works, right?

**Comment by famod** [comment]
Sure, just a funny coincidence that assertj started making some trouble shortly after you mentioned it.

**Comment by rsvoboda** [comment]
Hi. I'm not sure this change is a good one for the average John Doe developer who appreciates one BOM file which handles the dependencies for him.

https://mvnrepository.com/artifact/org.assertj/assertj-core is popular - Used By 12,504 artifacts
Assertj is a defacto standard library for testing and I think that its removal from Quarkus BOM will make some users unhappy. Ease of use and good developer experience is one of the themes for Quarkus.

Falko is power/super user of Quarkus and want's to control things probably more than the rest of the user base.
To me this change is to make happy power users. Implication on the common users should be considered too.

**Comment by gsmet** [comment]
@rsvoboda I don't think you have fully grasped the issue we have with Assertj: it breaks backward binary compatibility very often and it makes upgrades very hard because you end up with things compiled with previous version not working when we enforce the version with the BOM.
It happened already twice always for trivial changes they made.

So while I agree it was more convenient (that was the reason why I was against it at first), it's just not sustainable for us and the Platform members.

It has nothing to do with allowing things for power users given you can always override the version if you want.

**Comment by rsvoboda** [comment]
Hey. I was not aware of the troubles with Assertj and backward binary compatibility.
I wanted to just stress out the convenience for common users.
With the new fact in mind, I think it's good to proceed with the PR.
Sorry for the previous comment, should have checked all the details.

**Comment by famod** [comment]
Yeah, the (my) initial motivation was different (and debatable), but now it's more urgent.

@gsmet @aloubyansky are we waiting for anything to merge this? `main` is already for 2.8, so...

**Comment by aloubyansky** [comment]
@gsmet needs to approve it.

**Review by gsmet — APPROVED** [review]

**Pull request merged**
